[Autograd] Cond Higher-Order Operation

bohnstingl opened the pull request:
This is an updated PR to equip cond with the autograd feature and replaces the old [PR](https://github.com/pytorch/pytorch/pull/126007)

@ydwu4 I tried to incorporate your requests already.

Currently there are two problems that I struggle with solving:

1. There seems to be an import issue when trying to import cond in `torch/__init__.py`, see [here](https://github.com/bohnstingl/pytorch/blob/8a704035c964acfff35e5b0a2f0acc050895e3eb/torch/__init__.py#L1914-L1916). Therefore, I had to comment those lines, which resolved the import issues, but I believe cond is not proberly exposed as torch.cond.
2. I am not entirely sure how to deal with the opinfo test in `hop_db.py`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bohnstingl committed:
Changes required to be able to build PyTorch in conda environment with CUDA

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into cond_autograd

bohnstingl committed:
WIP: Cond Autograd

bohnstingl committed:
WIP: cond autograd forward

bohnstingl committed:
Fixed cond autograd

bohnstingl committed:
Executed 'test_control_flow.py' and 'test_higher_order_ops.py'

bohnstingl committed:
Added additional testcases

ydwu4 commented on functorch/experimental/control_flow.py:
```diff
@@ -1,8 +1,10 @@
-from torch import cond  # noqa: F401
+# TODO: Currently cond is not fully exposed under torch, e.g. torch.cond, but it causes import errors 
```
oh, this might be because of an circular import. We can break tie in cond.py by moving 
```
from torch._functorch.aot_autograd import AOTConfig, create_joint, from_fun
``` 
into `create_fw_bw_graph` and leave these imports unmodified.

bohnstingl replied:
I have moved the import into the `create_fw_bw_graph` as you suggested and this resolved the circular import issue

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -256,6 +256,109 @@ def false_fn(x):
         pred = torch.tensor(False, device="cuda")
         result = cond(pred, true_fn, false_fn, [x])
         self.assertEqual(result, torch.cos(x))
+        
```
For each of these added tests, can you add the following test to check the graph:
```python
# Given following f
def f(...):
   result = forward_with_cond
   grad_out = torch.ones_like(result)
   return torch.autograd.grad(result, (x,), grad_out)

# can use make_fx to capture the graph
gm = make_fx(f)(...)
# check the graph
self.assertExpectedInline(gm.print_readable().strip(), """hahaha""")
```
You can use `EXPECTTEST_ACCEPT=1 python test/functorch/test_control_flow.py` to automatically populate the expected inline string.

bohnstingl replied:
I added the graph checks as you suggested.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -476,6 +579,52 @@ def f(x, y):
         graph = make_fx(f, tracing_mode="symbolic")(x, torch.tensor(False))
         self.assertEqual(graph(x, torch.tensor(True)), f(x, torch.tensor(True)))
 
+    @skipIfTorchDynamo("Graph is not captured by backend if test with dynamo")
```
This test doesn't check backward graph. We can use the way that's mentioned in previous comment to test the graph.

bohnstingl replied:
This test still fails, as the graph module does not contain the backward graph. I have realized that the root cause of this is that in the [forward function of CondAutogradOp](https://github.com/pytorch/pytorch/pull/126911/files#diff-4093d24068320f82997aecfaf4260aa495e3b4322e178cb5625f9555e3f8b0f5R359-R363), the `cond_op` is torch.compiled.

bohnstingl replied:
This test is now working as expected again

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -2358,7 +2487,8 @@ def test_cond_vmap(self, predType, innerFnType, nOperands, nClosure, nesting):
     def test_cond_vmap_simple(self):
         def fn(x):
-            return torch.cond(
+            # return torch.cond(
```
By removing the circular imports, we should be able to leave these unchanged.

bohnstingl replied:
Indeed this resolved the issue

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -143,13 +147,186 @@ def _validate_input(pred, true_fn, false_fn, operands):
+
+    # Note:[HOP create fw_bw graph] We create "clean" environments for make_fx by suspending all dispatch keys
+    # between Autograd and Python key. Currently, we only suspend functionalization but more can be
+    # added when required. Will encounter two problems if we don't suspend functionalization:
```
This redundancy with map is not ideal. We should probably think about ways of unifying the implementation into a helper function and use the same helper function for control flow ops.

bohnstingl replied:
This test still fails, as the graph module does not contain the backward graph. I have realized that the root cause of this is that in the [forward function of CondAutogradOp](https://github.com/pytorch/pytorch/pull/126911/files#diff-4093d24068320f82997aecfaf4260aa495e3b4322e178cb5625f9555e3f8b0f5R359-R363), the `cond_op` is torch.compiled.

bohnstingl replied:
I consolidated some of the commonly used methods and put them into the [utils.py](https://github.com/pytorch/pytorch/pull/126911/files#diff-94838db6991b14cfeec36d4eb0e0ff6fadc08624a7c3ac967820f7f7ae267e70R205-R300)

ydwu4 commented:
Please also do a rebase and address the conflicts

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into cond_autograd

bohnstingl committed:
Fixed import issues
Merged with main Pytorch main branch
Added new testcases
Consolidated common functions for the create_fw_bw_graph with map.py

bohnstingl committed:
Fixed test_cond_vmap_single_input_with_closure testcase

bohnstingl commented:
@ydwu4 Thank you very much for your comments and reviews. I updated the files accordingly. There are two open points though:

- [One test](https://github.com/pytorch/pytorch/blob/f6cfa9627d5eea922cab8faee6f92bf87c204217/test/functorch/test_control_flow.py#L2547) is failing now after the rebase and I can't quite figure out why.
- I had to comment out these [lines](https://github.com/pytorch/pytorch/blob/9617494cae26e66550adfa77ca7eb10f16488fd3/test/functorch/test_control_flow.py#L2807-L2810) in order to make the test work again. Is this acceptable, or how can one resolve this issue?

ydwu4 reviewed (commented):
Good work overall! We're close to land the feature. Left some comments on the pr. Mostly about:
1. more tests, see the comments for some ideas of testing. Could use your imagination to create more tests.
2. we should not call torch.compile again when we're dispatching autograd key. We should just call the operator itself.

bohnstingl committed:
Introduced lower level function for create_fw_bw_graph
Fixed test cases
Removed unneccessary compil step in AutoGrad

bohnstingl committed:
Updated low-level create_fw_bw_graph function (compatible with while_loop)
Fixed issues with cond
Added several additional testcases

bohnstingl committed:
Cleanup

bohnstingl commented on test/functorch/test_control_flow.py:
```diff
@@ -256,6 +256,289 @@ def false_fn(x):
+        # can use make_fx to capture the graph
+        gm = make_fx(f)(pred, x)
+        # check the graph
+        # self.assertExpectedInline(gm.print_readable().strip(), """hahaha""")
```
I used this method for all testcases, except for [one](https://github.com/pytorch/pytorch/blob/40d584c06e151defbb22ade1067e57c4a720f5ab/test/functorch/test_control_flow.py#L416) where the output would have been too large.

bohnstingl commented on test/functorch/test_control_flow.py:
```diff
@@ -256,6 +256,289 @@ def false_fn(x):
+    )
+            
+    def test_cond_autograd_complex(self):
+        def true_fn(x):
```
I added those [test cases](https://github.com/pytorch/pytorch/blob/40d584c06e151defbb22ade1067e57c4a720f5ab/test/functorch/test_control_flow.py#L333-L1375)

bohnstingl commented on test/functorch/test_control_flow.py:
```diff
@@ -256,6 +256,289 @@ def false_fn(x):
+    )
+        
+    def test_cond_autograd_inner_tensor(self):
+        def true_fn(x):
```
I added those test cases along with some more. 

However:

1. In the case of the pytree output, the gradient computation becomes tricky. Therefore, for this case, I just checked the [forward path](https://github.com/pytorch/pytorch/blob/40d584c06e151defbb22ade1067e57c4a720f5ab/test/functorch/test_control_flow.py#L670).
2. I did not consolidate the used models in any special place. Please advice if that is needed.

bohnstingl commented on test/functorch/test_control_flow.py:
```diff
@@ -256,6 +256,289 @@ def false_fn(x):
+            grad_out = torch.ones_like(result)
+            return torch.autograd.grad(result, (x,), grad_out)
+
+        # can use make_fx to capture the graph
```
Done

bohnstingl commented on torch/_higher_order_ops/cond.py:
```diff
@@ -243,10 +346,36 @@ def cond_op_dense(pred, true_fn, false_fn, operands):
+        
+        with torch._C._AutoDispatchBelowAutograd():
+            with _set_compilation_env(), torch._dynamo.utils.disable_cache_limit():
+                return torch.compile(cond_op, backend="eager", fullgraph=True)(
```
I removed the additional compilation step as it was indeed not necessary. Thanks for the explanation.

bohnstingl commented on torch/_higher_order_ops/cond.py:
```diff
@@ -243,10 +346,36 @@ def cond_op_dense(pred, true_fn, false_fn, operands):
+
+@cond_op.py_impl(DispatchKey.Autograd)
+def cond_autograd(pred, true_fn, false_fn, operands):
+    num_mapped_args = len(operands)
```
Done

bohnstingl commented on torch/_higher_order_ops/cond.py:
```diff
@@ -142,12 +146,111 @@ def _validate_input(pred, true_fn, false_fn, operands):
+        keep_inference_input_mutations=False,
+    )
+
+    # Note:[HOP create fw_bw graph] We create "clean" environments for make_fx by suspending all dispatch keys
```
Done

bohnstingl commented on torch/_higher_order_ops/cond.py:
```diff
@@ -142,12 +146,111 @@ def _validate_input(pred, true_fn, false_fn, operands):
+    with suspend_functionalization(), disable_functional_mode():
+        with disable_proxy_modes_tracing():
+
+            num_mapped_args = len(operands)
```
Done

bohnstingl commented on torch/_higher_order_ops/cond.py:
```diff
@@ -142,12 +146,111 @@ def _validate_input(pred, true_fn, false_fn, operands):
+            unwrapped_mapped_operands = pytree.tree_map(_from_fun, operands)
+            example_operands = unwrapped_mapped_operands
+
+            #Note, the true_fn and the false_fn produce the same output
```
Done

bohnstingl commented on torch/_higher_order_ops/cond.py:
```diff
@@ -142,12 +146,111 @@ def _validate_input(pred, true_fn, false_fn, operands):
 cond_op = HigherOrderOperator("cond")
+cond_op.__module__ = "torch.ops.higher_order"
+
+def create_fw_bw_graph(true_fn, false_fn, *operands):
```
I created such a low-level function as suggested and it makes sense to me. I also integrated this low-level function into the [while loop](https://github.com/pytorch/pytorch/blob/8008d5dbfc9588aa27aa1a5380ae5c51b3b57900/torch/_higher_order_ops/while_loop.py#L96) in this [PR](https://github.com/pytorch/pytorch/pull/124573).

However, there is one problem with that I realized: The `create_fw_bw_graph` function of the `cond`, or any other higher_order_op is invoked twice. Once in the forward pass, which expected. However, it is invoked a second time during the backward pass, which causes some trouble. In fact, I had to add a "detection mechanism" for that case and avoid creating a `joint_fn` of an already created `joint_fn`. If this second invocation can be removed, it would simplify things quite a bit.

bohnstingl replied:
@ydwu4 What do you think about this lower-level function in `utils.py`?

As I mentioned, I still see this problem that the `create_fw_bw_graph_branches` function is invoked in the forward and in the backward path. For `cond` it might not be too much of a problem, but for other operations, such as `while_loop`, it may be problematic. 

This is because in the case of the while_loop for example, the cond and the body function in the backward path have slightly different signatures compared to the forward path. 
```
def cond_fn(x):
    return x.sum() < 10
```
vs.
```
def bw_cond_fn(*args):
    return len(args[-1]) > 1
```
where `args` is a tuple of `(grads, fw_outputs)`



For the first invokation of `create_fw_bw_graph_branches` (in the forward path) this is fine, but it causes problems when this function is called again in the backward path. One has to explicitly detect that one is in the backward path and avoid calling the lower-level function, but rather just return dummy graphs, as they are not used anyways. [Here](https://github.com/pytorch/pytorch/blob/b7aefd85743e51b708fcb6b61e43fb330f7ee25d/torch/_higher_order_ops/while_loop.py#L97-L105) is an example of what I mean.
Do you have any thoughts on this?

ydwu4 replied:
We can make fw_outputs an argument instead of computing it inside of the interface. https://github.com/pytorch/pytorch/pull/126911/files#diff-94838db6991b14cfeec36d4eb0e0ff6fadc08624a7c3ac967820f7f7ae267e70R294 

BTW, let's pause while_loop and bit and I'll need some time to think this through.

bohnstingl replied:
Makes sense to me, I have updated the [low-level function](https://github.com/pytorch/pytorch/blob/630549a7e12c32b0f0f9a366c53e247b9a32a230/torch/_higher_order_ops/utils.py#L266) accordingly.

I have already paused the development on the while_loop as you suggested. However, I just tried to have it in mind for the `create_fw_bw_graph` function. In general, I am quite eager to have the while_loop feature ready, so whenever there is something that I can help, please do let me know.

bohnstingl commented on torch/_higher_order_ops/cond.py:
```diff
@@ -142,12 +146,111 @@ def _validate_input(pred, true_fn, false_fn, operands):
+
+            return pytree.tree_map(maybe_clone, grads)
+        
+        def joint_f_false(*joint_mapped_args):
```
Indeed, the redundancy has been alleviated by utilizing the new low-level API. However, there is a problem with this new API, see comment above.

bohnstingl commented on test/functorch/test_control_flow.py:
```diff
@@ -2439,10 +2804,10 @@ def fn(x):
-            res = torch.vmap(fn, in_dims=(0,))(
-                a,
-            )
+        # with unittest.mock.patch("torch._dynamo.config.error_on_recompile", True):
```
With the latest changes, the commented section could be reverted.

bohnstingl commented on test/functorch/test_control_flow.py:
```diff
@@ -1804,26 +2189,6 @@ def f(xs, y):
         ):
             functional_f(*example_inputs)
 
-    def test_cond_autograd_fail(self):
```
I have changed the test name and modified it. Now, instead of checking that cond_autgrad raises an error, it checks whether this is NOT the case.

I also added a call to torch.compile(f), which works as expected.

bohnstingl commented on functorch/experimental/control_flow.py:
```diff
@@ -5,4 +5,4 @@
     _unstack_pytree,
     map,
-)
+)
```
Done

bohnstingl commented:
I tried to incorporate the review comments from above. While creating additional testcases, I stumbled over a couple of problems:

1. As mentioned above, the `create_fw_bw_graph` function is invoked twice, which is somewhat problematic.
2. Some `nn.Modules` caused problems when used in the `true_fn/false_fn`. For example, when the nn.Linear or the nn.GRU module is used, the check in the trace fails, as the `true_out` or the false_out in that case do not have meta data. Therefore, I [extended](https://github.com/pytorch/pytorch/blob/40d584c06e151defbb22ade1067e57c4a720f5ab/torch/_higher_order_ops/cond.py#L295-L299) this check to avoid this issue.
3. The testcase `TestControlFlowTraced.test_cond_make_fx_preserve_stack_trace_for_nodes_in_subgraph` is failing and I don't quite know why. The respective dictionary does not track all the meta data correctly. For example, the target output is `[('cond', <torch._ops.HigherOrderOperator object at 0x7f5b4d710810>), ('cos', 'cos')]`, while the produced output is `[('cond', <torch._ops.HigherOrderOperator object at 0x7f5b4d710810>)]`. Here is the stack trace:
```
======================================================================
FAIL: test_cond_make_fx_preserve_stack_trace_for_nodes_in_subgraph (__main__.TestControlFlowTraced.test_cond_make_fx_preserve_stack_trace_for_nodes_in_subgraph)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data_malta3_ssd/pytorch/torch/testing/_internal/common_utils.py", line 2756, in wrapper
    method(*args, **kwargs)
  File "/data_malta3_ssd/pytorch/test/functorch/test_control_flow.py", line 3413, in test_cond_make_fx_preserve_stack_trace_for_nodes_in_subgraph
    self.assertEqual(all_source_fns, new_source_fns)
  File "/data_malta3_ssd/pytorch/torch/testing/_internal/common_utils.py", line 3607, in assertEqual
    error_metas = not_close_error_metas(
                  ^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch/torch/testing/_comparison.py", line 1220, in not_close_error_metas
    raise error_meta.to_error() from None  # noqa: RSE102
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: The length of the sequences mismatch: 2 != 1

The failure occurred for item [0]

To execute this test, run the following from the base repo dir:
     python test/functorch/test_control_flow.py -k TestControlFlowTraced.test_cond_make_fx_preserve_stack_trace_for_nodes_in_subgraph
```

bohnstingl committed:
Added some more test cases with loops

bohnstingl committed:
Cleaned utils.py

bohnstingl committed:
Fixed lintrunner and some of the cond testcases

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -183,7 +226,16 @@ def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
+        # is composed of the multiplication, then the requires_grad attribute is
+        # set to False and thus the tracing of the cond errors out.
+        if (
+            true_out is not None
```
If one branch return None but the other branch is not returning None, we should also error out. Can try to fix the error message to include this case.  We can have a expectedRaise test for this corner case.

bohnstingl replied:
I added an additional [testcase](https://github.com/pytorch/pytorch/blob/630549a7e12c32b0f0f9a366c53e247b9a32a230/test/functorch/test_control_flow.py#L799) for this behavior

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -183,7 +226,16 @@ def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
         true_out = flat_true_outs[i]
         false_out = flat_false_outs[i]
-        if true_out.meta["tensor_meta"] != false_out.meta["tensor_meta"]:
+        # TODO: If a torch nn module such as Linear or GRUCell is used, then the
```
Is this TODO still valid?

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -183,7 +226,16 @@ def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
-        if true_out.meta["tensor_meta"] != false_out.meta["tensor_meta"]:
+        # TODO: If a torch nn module such as Linear or GRUCell is used, then the
+        # meta data of the output is None and cannot be compared
+        # TODO: If inside the dictionary, inside the list, the first element
```
We should figure out why this happens and fix it.

bohnstingl commented on test/functorch/test_control_flow.py:
```diff
@@ -257,6 +257,1372 @@ def false_fn(x):
+            return torch.autograd.grad(x_new, (x,), grad_out)
+
+        # TODO: During compilation, the metadata of the true_fn has the 
+        # requires_grad attribute set to False, while the false_fn has it
```
Very good catch @ydwu4, I think this might even be the root cause of some of the other issues that we see. In this particular case, the `make_fx` is introducing this detach operation even in the forward path. So in case of
```
def forward(self, l_args_3_0_):
    l_args_3_0__1 = l_args_3_0_
    relu = torch.nn.functional.relu(l_args_3_0__1, inplace = False);  l_args_3_0__1 = None
    return (relu,)
```
`make_fx` turns this into
```
def forward(self, arg0_1):
    relu = torch.ops.aten.relu.default(arg0_1);  arg0_1 = None
    detach = torch.ops.aten.detach.default(relu)
    return (relu,)
```
I don't quite know what is the reason for this and I am not familiar with the inner workings of `make_fx`
This is the same for `torch.nn.GRUCell`. However, this does not happen if I just create a custom nn.Module that itself implements the ReLU function
```
class SimpleNN(torch.nn.Module):
    def __init__(self):
        super().__init__()
    def forward(self, x):
        return torch.maximum(torch.zeros_like(x), x)
nn_module_false = SimpleNN()

def false_fn(x):
    return nn_module_false(x)
```
which when make_fx is applied gives
```
def forward(self, arg0_1):
    zeros_like = torch.ops.aten.zeros_like.default(arg0_1, pin_memory = False)
    maximum = torch.ops.aten.maximum.default(zeros_like, arg0_1);  zeros_like = arg0_1 = None
    return (maximum,)
```
Do you have any hints on how to approach this?

ydwu4 replied:
yeah, I can take a look. For a normal function like below,
```python
import torch
from torch.fx.experimental.proxy_tensor import make_fx

def f(x):
    return torch.nn.functional.relu(x)

gm = make_fx(f)(torch.randn(3, 2))
```
we'll get a normal graph as shown below:
```python
class f(torch.nn.Module):
    def forward(self, x_1: "f32[3, 2]"):
        # No stacktrace found for following nodes
        relu: "f32[3, 2]" = torch.ops.aten.relu.default(x_1);  x_1 = None
        return relu
```
So the bug is probably related to cond.

bohnstingl replied:
Thank you very much for looking into this!
I will continue to think about more such creative usecases, that potentially unveils bugs :-)

Some information that might help you in this issue: I tried to test several components of `torch.nn` and it seems that this issues is not present for all of them. For example, the detach operations `are not` present for 
- `torch.nn.Identity()`
- `torch.nn.Dropout()`
- `torch.nn.Linear()`
- `torch.nn.functional.threshold`
- `torch.nn.functional.glu`
- `torch.nn.functional.gelu`
- `torch.nn.functional.leaky_relu`
- `torch.nn.functional.logsigmoid`
- `torch.nn.functional.softplus`
- `torch.nn.functional.hardtanh`
- `torch.nn.functional.elu`
- `torch.nn.functional.selu`
- `torch.nn.functional.rrelu`

but, `are` present for
- `torch.nn.RNN`
- `torch.nn.LSTM`
- `torch.nn.GRU`
- `torch.nn.Transformer`
- `torch.nn.functional.relu`
- `torch.nn.functional.softmax`
- `torch.nn.functional.tanh`
- `torch.nn.functional.softmin`

At first, there was no clear pattern to me, but after quite some investigations into the failing cases, it seems that this behavior is mostly triggered by the `tanh`, `relu` or `softmax` functions. At least, all the failing cases I could track back to one of these functions. I hope this helps.

ydwu4 replied:
Thanks for the thorough study!  @zou3519 told me when autograd saves a Tensor for backwards, there's a call to .detach() and pytorch has auto codegen for autograd: see `derivatives.yaml`. Generated code is in  `VariableTypeEverything.cpp`. In the case of `relu`, the result is saved like this:     `grad_fn->result_ = SavedVariable(result, true);`

bohnstingl replied:
Ah, okay. Then this at least explains the behavior of what we see. To me it is interesting though why the `rrelu` or the `leaky_relu` functions do not show this behavir? Do these functions not save any tensor for the backward path?

ydwu4 replied:
Good question. Not sure about the detail. Checked leaky_relu a bit: the difference between relu and leaky_relu in the generated code is that it doesn't store the SavedVariable in grad_fn like relu did. However, from derivatives.yarml, I cannot tell why we have such diference. They all return `auto_element_wise`. Might have other places of controlling the code-gen logic. This is a separate discussion though (or is this necessary for getting correct pytree output?).

bohnstingl replied:
Okay, thank you.

I think for the moment, we could just leave a TODO, but we need to come back eventually to this, I guess. I also observed that a similar (potentially related) issue with the `requires_grad` flag occurs for the testcase [test_cond_in_forloop](https://github.com/pytorch/pytorch/blob/630549a7e12c32b0f0f9a366c53e247b9a32a230/test/functorch/test_control_flow.py#L678), for example. In that case, there is no `relu` function involved, but still the `true_out.meta["tensor_meta"]` has `requires_grad=False`, while `true_out.meta["tensor_meta"]` has set it to True

ydwu4 replied:
Can you check how `tensor_meta` is created and why it is set to False in one branch?

bohnstingl replied:
I tried to go down the rabbit hole and actually found something puzzling to me. In the simple case, 
```
def true_fn(x):
    return x.sin()
def false_fn(x):
    return x.cos()
```
The metadata are
```
true_out.meta["tensor_meta"] = TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={}))
false_out.meta["tensor_meta"] = TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={}))
```
However, if in one branch the input arguments are just passed through, e.g.,
```
def true_fn(x):
    return x
def false_fn(x):
    return x.cos()
```
the metadata become
```
true_out.meta["tensor_meta"] = TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=True, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={}))
false_out.meta["tensor_meta"] = TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={}))
```

If one applies even a simple multiplication with 1 to the input argument, the metadata match again
```
def true_fn(x):
    return x*1
def false_fn(x):
    return x.cos()
true_out.meta["tensor_meta"] = TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={}))
false_out.meta["tensor_meta"] = TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={}))
```
To me this behavior is unexpected, but I don't know where this comes from.

bohnstingl replied:
@ydwu4. Do you maybe have any insights into this? Thank you

bohnstingl committed:
Updated low-level create_fw_bw_graph function

bohnstingl committed:
Updated low-level create_fw_bw_graph function
Fixed and added further testcases

bohnstingl commented on test/functorch/test_control_flow.py:
```diff
@@ -1818,11 +3249,14 @@ def f(x, y):
 
         # Ensure no error is thrown when not running backward
-        f(*example_inputs)
+        res = torch.compile(f)(*example_inputs)
```
I added this evaluation

bohnstingl commented on test/functorch/test_control_flow.py:
```diff
@@ -257,6 +257,1372 @@ def false_fn(x):
+            result = cond(pred, true_fn, false_fn, ({"t": [a, {"b": b}, (c,)]},))
+            self.assertEqual(result, fn({"t": [a, {"b": b}, (c,)]}))
+
+            # TODO: Gradient computation for such complex pytree output does not work
```
What I am observing is that the following errors out:

```
def false_fn(x):
    return {"res": [x["t"][1]["b"], (x["t"][2][0],)]}

result_exp = fn({"t": [a, {"b": b}, (c,)]})
grad_out = torch.ones_like(a)
expected_grads = torch.autograd.grad(result_exp, (a,), grad_out)
```
with the Exception:
```
Traceback (most recent call last):
  File "/data_malta3_ssd/pytorch/torch/testing/_internal/common_utils.py", line 2756, in wrapper
    method(*args, **kwargs)
  File "/data_malta3_ssd/pytorch/test/functorch/test_control_flow.py", line 925, in test_cond_autograd_same_pytree_output
    expected_grads = torch.autograd.grad(result_exp, (a,), grad_out)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch/torch/autograd/__init__.py", line 402, in grad
    grad_outputs_ = _make_grads(
                    ^^^^^^^^^^^^
  File "/data_malta3_ssd/pytorch/torch/autograd/__init__.py", line 91, in _make_grads
    if out.is_nested or first_grad.is_nested:
       ^^^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'is_nested'
```
So I don't think this has anything to do with cond, it just doesn't work in this setting even for regular autograd. The respective testcase that fails is [here](https://github.com/pytorch/pytorch/blob/630549a7e12c32b0f0f9a366c53e247b9a32a230/test/functorch/test_control_flow.py#L945).

ydwu4 replied:
Oh, yeah, I forgot to mention that we need to flatten the result_exp and grad_out to make torch.autograd.grad work [doc is here](https://pytorch.org/docs/stable/generated/torch.autograd.grad.html).

ydwu4 commented:
I'm back! do you mind rebasing on main?

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into cond_autograd

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into cond_autograd

bohnstingl commented:
I did a rebase, but I currently have trouble compiling from main. I ensured that all submodules are also up-to-date, but I get the following error
```
/pytorch/aten/src/ATen/cpu/Utils.cpp: In function 'bool at::cpu::is_cpu_support_amx_tile()':
/pytorch/aten/src/ATen/cpu/Utils.cpp:37:34: error: 'cpuinfo_has_x86_amx_tile' was not declared in this scope; did you mean 'cpuinfo_has_x86_mmx_plus'?
   37 |   return cpuinfo_initialize() && cpuinfo_has_x86_amx_tile();
      |                                  ^~~~~~~~~~~~~~~~~~~~~~~~
      |                                  cpuinfo_has_x86_mmx_plus
```
Have you every seen this error and potentially know how to resolve it? Maybe you have more luck

bohnstingl committed:
Merge branch 'pytorch:main' into cond_autograd

bohnstingl committed:
Updated testcases for cond

bohnstingl committed:
Merge branch 'cond_autograd' of github.com:bohnstingl/pytorch into cond_autograd

ydwu4 commented:
>  but I currently have trouble compiling from main.

Yean, are the third-party library changes and cmake changes necessary? Can we just keep the changes in _higher_order_ops and tests? I'll cherry-pick some of the  changes and do some debugging for the problem you mentioned.

bohnstingl commented:
It works again. If you do a fresh checkout from my branch, PT can be compiled from source again as normal.

ydwu4 committed:
cond autograd

[ghstack-poisoned]

bohnstingl committed:
Integrated additions from https://github.com/pytorch/pytorch/pull/130607

bohnstingl committed:
Updated cond.py

bohnstingl committed:
Updated cpuinfo to most recent version

bohnstingl committed:
Reverted cpuinfo version

bohnstingl committed:
Updated third_party modules to pytorch:main version

bohnstingl committed:
Merge branch 'cond_autograd' of github.com:bohnstingl/pytorch into HEAD

bohnstingl committed:
Merge branch 'gh/ydwu4/132/base' of github.com:pytorch/pytorch into cond_autograd

bohnstingl committed:
Merge branch 'cond_autograd' of github.com:bohnstingl/pytorch into cond_autograd

ydwu4 committed:
Update on "[NOT FOR REVIEW] cond autograd."


A clean build for https://github.com/pytorch/pytorch/pull/126911 and fix most test failures.



[ghstack-poisoned]

ydwu4 commented:
Updated https://github.com/pytorch/pytorch/pull/130607 to address remaining test failures. When the ci passes for #130607, we can start to migrate the changes here then merge the PR.

ydwu4 committed:
Update on "[NOT FOR REVIEW] cond autograd."


A clean build for https://github.com/pytorch/pytorch/pull/126911 and fix most test failures.



[ghstack-poisoned]

ydwu4 committed:
Update base for Update on "[NOT FOR REVIEW] cond autograd."


A clean build for https://github.com/pytorch/pytorch/pull/126911 and fix most test failures.



[ghstack-poisoned]

ydwu4 committed:
Update on "[NOT FOR REVIEW] cond autograd."


A clean build for https://github.com/pytorch/pytorch/pull/126911 and fix most test failures.



[ghstack-poisoned]

ydwu4 committed:
Update base for Update on "[NOT FOR REVIEW] cond autograd."


A clean build for https://github.com/pytorch/pytorch/pull/126911 and fix most test failures.



[ghstack-poisoned]

ydwu4 committed:
Update base for Update on "[NOT FOR REVIEW] cond autograd."


A clean build for https://github.com/pytorch/pytorch/pull/126911 and fix most test failures.



[ghstack-poisoned]

ydwu4 committed:
Update on "[NOT FOR REVIEW] cond autograd."


A clean build for https://github.com/pytorch/pytorch/pull/126911 and fix most test failures.



[ghstack-poisoned]

bohnstingl committed:
Merge branch 'gh/ydwu4/132/head' of github.com:pytorch/pytorch into cond_autograd

bohnstingl committed:
Merge branch 'gh/ydwu4/132/base' of github.com:pytorch/pytorch into cond_autograd

bohnstingl committed:
Fixed some more testcases

bohnstingl commented on test/functorch/test_control_flow.py:
```diff
@@ -259,6 +262,1063 @@ def false_fn(x):
+    return {'res': [view, (view_1,)]}""",  # noqa: B950
+        )
+
+    def test_cond_autograd_torch_nn_module(self):
```
This testcase still shows detach operations for certain operations, for example torch.nn.functional.relu. Is this something that we need to worry about?

ydwu4 replied:
Yeah, detach is expected behavior. This happens when "autograd saves a Tensor for backwards". What tensor to be saved varies from each operators to another.

bohnstingl commented:
Hi @ydwu4,
Thank you very much for your effort! I have integrated your changes from https://github.com/pytorch/pytorch/pull/130607 and reverted the changes to the third_party modules, as well as the changes to the cmake file.
I did another round of fixes of the testcases, but there is one test in `TestControlFlowTraced`, namely [test_cond_make_fx_preserve_stack_trace_for_nodes_in_subgraph](https://github.com/pytorch/pytorch/blob/8d03c22a81023f18034e47ed75fd540619163012/test/functorch/test_control_flow.py#L3443) that fails.
This failure was present already from the very beginning of this PR and I mentioned it already [here](https://github.com/pytorch/pytorch/pull/126911#issuecomment-2131603193). It appears that the graph created with `make_fx` does not track the actual `true_fn` or `false_fn`. For example, when using `gm, _ = torch._dynamo.export(foo)(inp)` the source_fn_stack looks like
```
source_fn_stack
[('cond', <torch._ops.HigherOrderOperator object at 0x7f04d20bba50>), ('sin', 'sin')]
stack_trace
  File "/data_malta3_ssd/pytorch/test/functorch/test_control_flow.py", line 3436, in foo
    return cond(x.shape[0] == 4, true_fn, false_fn, (x,))
  File "/data_malta3_ssd/pytorch/torch/_higher_order_ops/cond.py", line 112, in cond
    return cond_op(pred, true_fn, false_fn, operands)
  File "/data_malta3_ssd/pytorch/test/functorch/test_control_flow.py", line 3433, in false_fn
    return x * x.sin()
```
However, when using `new_gm = make_fx(run_with_interpreter)(inp)`, the source_fn_stack is only 
```
source_fn_stack
[('cond', <torch._ops.HigherOrderOperator object at 0x7f04d20bba50>)]
stack_trace
  File "/data_malta3_ssd/pytorch/test/functorch/test_control_flow.py", line 3436, in foo
    return cond(x.shape[0] == 4, true_fn, false_fn, (x,))
  File "/data_malta3_ssd/pytorch/torch/_higher_order_ops/cond.py", line 112, in cond
    return cond_op(pred, true_fn, false_fn, operands)
```

Furthermore, there still exists the issue that some operations, such as torch.nn.functional.relu, introduces detach operations in the graph. This is following this discussion [here](https://github.com/pytorch/pytorch/pull/126911#discussion_r1623175699) and this [testcase](https://github.com/pytorch/pytorch/blob/8d03c22a81023f18034e47ed75fd540619163012/test/functorch/test_control_flow.py#L935). Is this something that we need to worry about at the moment?

ydwu4 committed:
Update base for Update on "[NOT FOR REVIEW] cond autograd."


A clean build for https://github.com/pytorch/pytorch/pull/126911 and fix most test failures.



[ghstack-poisoned]

ydwu4 committed:
Update on "[NOT FOR REVIEW] cond autograd."


A clean build for https://github.com/pytorch/pytorch/pull/126911 and fix most test failures.



[ghstack-poisoned]

ydwu4 commented:
> when using new_gm = make_fx(run_with_interpreter)(inp), the source_fn_stack is only

This one should be fixed on the recent update.

>  introduces detach operations in the graph

This is expected behavior. We don't need to worry about (at least in this PR).

bohnstingl committed:
Merge branch 'gh/ydwu4/132/base' of github.com:pytorch/pytorch into cond_autograd

bohnstingl committed:
Merge branch 'gh/ydwu4/132/head' of github.com:pytorch/pytorch into cond_autograd

bohnstingl commented:
Great, thank you. I have pulled the latest changes and the failed testcase indeed is resolved. I hope that there are no further issues popping up

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -259,6 +259,877 @@ def false_fn(x):
+
+        # TODO: Once the while loop does support autograd, replace the manual while
+        # loop with the higher_order_operator
+        def while_loop_test(x):
```
Remove this test for now?

bohnstingl replied:
I removed the while-loop test for the moment

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -259,6 +259,877 @@ def false_fn(x):
+            grad_out = torch.ones_like(x_new)
+            return torch.autograd.grad(x_new, (x,), grad_out)
+
+        # TODO: During compilation, the metadata of the true_fn has the
```
We can remove the todo. This is expected behavior. We cannot safely test the requires_grad attribute at that key.

bohnstingl replied:
I removed it

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1927,32 +2863,15 @@ def f(x, y):
         )
-        # Due to x.shape[0] can be statically evaluated to be False, we can evaluate
-        # the backward.
+        # with self.assertRaisesRegex(RuntimeError, "Autograd not implemented for cond"):
```
Remove this

bohnstingl replied:
I removed it

ydwu4 reviewed (approved):
Left some more comments. After they're fixed, we're good to merge. Great job!

bohnstingl committed:
Final cleanup

bohnstingl commented:
I did a cleanup for the points that you mentioned

bohnstingl commented on torch/__init__.py:
```diff
@@ -2423,6 +2423,11 @@ def fn(model: _Callable[_InputT, _RetT]) -> _Callable[_InputT, _RetT]:
     )  # type: ignore[return-value]
 
 
+from torch import export as export
```
I reverted these specific changes

bohnstingl replied:
Unfortunately, with the latest revert the lint tests fail. If I run `lintrunner -a`, then I get back the same changes before the revert. I tried to push a new version which should be a tradeoff.

bohnstingl committed:
Fixes to lint test

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "[NOT FOR REVIEW] cond autograd."


A clean build for https://github.com/pytorch/pytorch/pull/126911 and fix most test failures.



[ghstack-poisoned]

ydwu4 committed:
Update base for Update on "[NOT FOR REVIEW] cond autograd."


A clean build for https://github.com/pytorch/pytorch/pull/126911 and fix most test failures.



[ghstack-poisoned]

ydwu4 committed:
Update on "[NOT FOR REVIEW] cond autograd."


A clean build for https://github.com/pytorch/pytorch/pull/126911 and fix most test failures.



[ghstack-poisoned]

ydwu4 committed:
Update base for Update on "[NOT FOR REVIEW] cond autograd."


A clean build for https://github.com/pytorch/pytorch/pull/126911 and fix most test failures.



[ghstack-poisoned]

bohnstingl committed:
Merge branch 'gh/ydwu4/132/head' of github.com:pytorch/pytorch into cond_autograd

ydwu4 commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "broke lint and functorch/test_aotdispatch https://hud.pytorch.org/pytorch/pytorch/commit/f7058b735e52a1d876912f8c96a594673a495007 Probably a landrace since both the test and lint passed on PR" -c landrace

ydwu4 commented:
@bohnstingl need a rebase, and import IS_WINDOWS in hop_db.py

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into cond_autograd

bohnstingl committed:
Skipping some cond tests in case of WINDOWS

bohnstingl commented:
I did a rebase to `main` and also added the test skipping for WINDOWS in `testing/hop_db.py`

bohnstingl commented:
Okay, there seems to be something wrong with the latest rebase. There are some tests failing that on my local machine work. For example 
```python test/export/test_export_training_ir_to_run_decomp.py -k TrainingIRToRunDecompExportTestExport.test_slice_with_floordiv_training_ir_to_decomp```
works for me. Moreover, there is one testcase failing which I think was introduced with the last rebase.

```
python test/functorch/test_aotdispatch.py -k TestEagerFusionOpInfoCPU.test_aot_autograd_symbolic_exhaustive_flex_attention_backward_simple_c
pu_float32
```

Returns an unexpected success for me locally.

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into cond_autograd

ydwu4 commented on torch/testing/_internal/hop_db.py:
```diff
@@ -196,7 +197,23 @@ def body_fn(iter_t, x):
+                "test_aot_autograd_exhaustive",
+                active_if=IS_WINDOWS,
+            ),
+            DecorateInfo(
```
Probably need to add another DecorateInfo for another windows test `PYTORCH_OPINFO_SAMPLE_INPUT_INDEX=0 python test\test_ops_gradients.py -k TestBwdGradientsCPU.test_fn_grad_cond_simple_cpu_float64` based on the CI signal. probably something like:
```
            DecorateInfo(
                unittest.expectedFailure,
                "TestBwdGradientsCPU",
                "test_fn_grad",
                active_if=IS_WINDOWS,
            ),
```
 Can you also add a `TODO (yidi): figure out window test failures`?

bohnstingl replied:
I added the DecorateInfo as well as a TODO

ydwu4 replied:
The DecorateInfo should be:
```
DecorateInfo(
    unittest.expectedFailure,
    "TestBwdGradients",
    "test_fn_grad",
    active_if=IS_WINDOWS,
)
```

bohnstingl replied:
I updated the DecorteInfo

bohnstingl committed:
Updated DecorateInfo for testcases for the case IS_WINDOWS

bohnstingl commented on torch/testing/_internal/hop_db.py:
```diff
@@ -240,6 +257,18 @@ def body_fn(iter_t, x):
             DecorateInfo(unittest.expectedFailure, "TestHOP", "test_pre_dispatch_export"),
             DecorateInfo(unittest.expectedFailure, "TestHOP", "test_serialize_export"),
             DecorateInfo(unittest.expectedFailure, "TestHOP", "test_retrace_export"),
+            DecorateInfo(
```
I removed it

bohnstingl commented on torch/testing/_internal/hop_db.py:
```diff
@@ -258,6 +287,18 @@ def body_fn(iter_t, x):
             DecorateInfo(unittest.expectedFailure, "TestHOP", "test_pre_dispatch_export"),
             DecorateInfo(unittest.expectedFailure, "TestHOP", "test_serialize_export"),
             DecorateInfo(unittest.expectedFailure, "TestHOP", "test_retrace_export"),
+            DecorateInfo(
```
I removed it

bohnstingl committed:
Updated DecorateInfo

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into cond_autograd

ydwu4 commented on torch/testing/_internal/hop_db.py:
```diff
@@ -91,13 +92,13 @@ def foo_impl_abstract(x, z):
 
 def simple_cond(x):
-    return torch.cond(x.shape[0] > 2, lambda x: x.cos(), lambda x: x.sin(), [x])
+    return torch.cond(x.sum() > 2, lambda x: x.cos(), lambda x: x.sin(), [x])
```
can you change the lambdas to return a tuple? e.g. (x.cos(), ). On windows, torch.compile is not compiling 
torch.cond causing errors. Then we can remove the two skip DecorateInfo starting from L203. We also need to expectFail test_convert_nn_module_with_nested_if_and_param due to the same reason.

bohnstingl replied:
Thank you for looking into this. I changed the return type to tuple and also removed the DecorateInfo. In addition, I tried to add an DecorateInfo for `test/export/test_converter.py`, for the testcase test_convert_nn_module_with_nested_if_and_param in 
`torch/testing/_internal/common_methods_invocations.py`, but I failed to do so. Could you maybe point me to the correct location where to add this DecorateInfo?

ydwu4 replied:
You could import IS_WIINDOWS (from torch.testing._internal.common_utils import IS_WINDOWS) and annotate the test case with:
```
    @unittest.skipIf(
        IS_WINDOWS,
        "torch.cond doesn't go through torch.compile on windows"
        "causing output not normalized as list",
    )
    def test_convert_nn_module_with_nested_if_and_param(self):
```

bohnstingl committed:
Fixed WINDOWS torch.compile issues with cond and added additional expectedFailure clause for Windows

bohnstingl committed:
Attempt to fix WINDOWS unittest issues

bohnstingl committed:
Another attempt

bohnstingl committed:
Revert TorchExort OpInfo

bohnstingl committed:
Annotated testcase directly to fail under WINDOWS

bohnstingl committed:
Fixed lintrunner issue

bohnstingl committed:
Added another skip annotation

XuehaiPan commented on torch/_higher_order_ops/cond.py:
```diff
@@ -253,9 +335,65 @@ def cond_op_dense(pred, true_fn, false_fn, operands):
+def cond_autograd(pred, true_fn, false_fn, operands):
+    # A shortcut for the case where all inputs don't require gradient,
+    # we skip tracing the forward and backward graph.
+    if all(
+        not t.requires_grad
+        for t in pytree.tree_flatten((pred, operands))[0]
+        if isinstance(t, torch.Tensor)
+    ):
```
```suggestion
    if pytree.tree_all_only(
        torch.Tensor,
        lambda t: not t.requires_grad,
        (pred, operands),
    ):
```

bohnstingl committed:
Update torch/_higher_order_ops/cond.py

Co-authored-by: Xuehai Pan <XuehaiPan@outlook.com>

bohnstingl commented:
@ydwu4, I think finally all unittests pass. Is there anything else that I can do?

ydwu4 commented:
@pytorchbot merge
